src/main/java/usi/si/seart/config/SchedulerConfig.java

```diff
@@ -47,4 +64,47 @@ public void handleError(@NotNull Throwable t) {
             log.error("Unhandled exception occurred while performing a scheduled job.", t);
         }
     }
+
+    /**
+     * Configuration for the thread pool responsible for code analysis and repository cloning.
+     *  No threads are instantiated when the pool is idle.
+     *  Idle threads die after 60 seconds.
+     *  The maximum pool size is configurable through the application.properties
+     */
+    @Bean(name = "GitCloning")
+    public Executor asyncExecutor() {
+        ThreadPoolTaskExecutor executor = new GitCloningThreadPoolExecutor();
+        executor.setCorePoolSize(0);
+        executor.setMaxPoolSize(maxPoolThreads);
+        executor.setQueueCapacity(10); // maximum number of tasks in the queue, after which more threads would be created
+        executor.setKeepAliveSeconds(60); // keep-alive time for idle threads
+        executor.setThreadNamePrefix("CloningThread");
+        executor.setRejectedExecutionHandler(new ThreadPoolExecutor.CallerRunsPolicy()); // policy to abort
+        executor.initialize();
+        return executor;
+    }
+
+    static class GitCloningThreadPoolExecutor extends ThreadPoolTaskExecutor {
+
+
+        /**
+         * Overrides the default cloning thread name generation.
+         * Allows for the reuse of thread names of dead cloning threads.
+         */
+        @NotNull
+        @Override
+        protected String nextThreadName() {
+            ThreadGroup currentGroup = Thread.currentThread().getThreadGroup();
+            int numThreads = currentGroup.activeCount();
+            Thread[] threads = new Thread[numThreads];
+            currentGroup.enumerate(threads);
+
+            return Stream.iterate(1, i -> i + 1)
+                    .map(i -> getThreadNamePrefix() + i)
+                    .filter(name -> Arrays.stream(threads).noneMatch(t -> Objects.equals(t.getName(), name)))
```

Comment:
Why stream inside a stream? Make a `Set` from the array of thread names, and then use the `contains` method.